File: lib/Model/touristQuizSession.dart

import '../Controller/request_controller.dart';

class touristQuizSession {

  int? tqsId;
  DateTime? startTime;
  DateTime? endTime;
  int? totalPoints;
  int? appUserId;

  touristQuizSession({
    required this.tqsId,
    required this.startTime,
    required this.endTime,
    required this.totalPoints,
    required this.appUserId,
  });

  touristQuizSession.getTqsId({
    required this.startTime,
    required this.endTime,
    required this.totalPoints,
    required this.appUserId,
  });

  touristQuizSession.fromJson(Map<String, dynamic> json)
      : tqsId = int.tryParse(json['tqsId'] ?? '') ?? 0,
        startTime = DateTime.parse(json['startTime'] as String? ??""),
        endTime = DateTime.parse(json['endTime'] as String? ??""),
        totalPoints = int.tryParse(json['totalPoints'] ?? '') ?? 0,
        appUserId = int.tryParse(json['appUserId'] ?? '') ?? 0;

  Map<String, dynamic> toJson() =>
      {
        'startTime': startTime?.toIso8601String(),
        'endTime': endTime?.toIso8601String(),
        'totalPoints': totalPoints,
        'appUserId': appUserId
      };

  Future<bool> submitQuizSession() async {
    RequestController req = RequestController(
        path: '/api/touristquizsession.php');

    // Set the body of the request
    req.setBody(toJson());

    // Make the HTTP request to submit the quiz session data
    await req.post();

    // Check the response status code
    if (req.status() == 200) {
      return true;
    } else {
      return false;
    }
  }

  Future<bool> getTqsId() async {
    RequestController req = RequestController(path: "/api/getTqsId.php");
    req.setBody(toJson());
    await req.post();
    if (req.status() == 200) {
      tqsId = int.tryParse(req.result()['tqsId'] ?? '');
      print(tqsId);
      return true;
    }
    else {
      return false;
    }
  }


}

File: PHP files/touristquizsession.php
<?php
// Uncomment the following lines if CORS headers are needed
// header("Access-Control-Allow-Origin: *");
// header("Access-Control-Allow-Methods: GET, POST, OPTIONS");
// header("Access-Control-Allow-Headers: *");
// header("Access-Control-Allow-Credentials: true");

$hostname = "localhost";
$database = "melakago";
$username = "root";
$password = "wingerssrc1519";

// Create a PDO instance for database connection
$db = new PDO("mysql:host=$hostname;dbname=$database", $username, $password);

// Set initial response code
http_response_code(404);
$response = new stdClass();

// Initialize $jsonbody variable using POST data
$jsonbody = json_decode(file_get_contents('php://input'));

if ($_SERVER["REQUEST_METHOD"] == "POST") {
    try {
        // Check if all required parameters are set
        if (
            isset($jsonbody->startTime) &&
            isset($jsonbody->endTime) &&
            isset($jsonbody->totalPoints) &&
            isset($jsonbody->appUserId)
        ) {
            // Insert the new quiz result
            $stmt = $db->prepare("INSERT INTO touristquizsession 
                (`startTime`, `endTime`, `totalPoints`, `appUserId`) 
                VALUES (:startTime, :endTime, :totalPoints, :appUserId)");

            $stmt->execute([
                ':startTime' => $jsonbody->startTime,
                ':endTime' => $jsonbody->endTime,
                ':totalPoints' => $jsonbody->totalPoints,
                ':appUserId' => $jsonbody->appUserId
            ]);

            http_response_code(200);
            $response->error = "Quiz result successfully registered";
        } else {
            http_response_code(400); // Bad Request
            $response->error = "Missing required parameters";
        }
    } catch (Exception $ee) {
        http_response_code(500);
        $response->error = "Error occurred " . $ee->getMessage();
    }
}

// Send the response as JSON
echo json_encode($response);
exit();
?>
